src/coreclr/tools/aot/ILCompiler.ReadyToRun/JitInterface/CorInfoImpl.ReadyToRun.cs

```diff
@@ -2604,5 +2603,123 @@ private void reportInliningDecision(CORINFO_METHOD_STRUCT_* inlinerHnd, CORINFO_
                 _inlinedMethods.Add(inlinee);
             }
         }
+
+        /// <summary>
+        /// Try to resolve a given virtual static interface method on a given constrained type and its base types.
+        /// </summary>
+        /// <param name="constrainedType">Type to attempt virtual static method resolution on</param>
+        /// <param name="interfaceType">Interface declaring the method</param>
+        /// <param name="interfaceMethod">Interface method to resolve</param>
+        /// <param name="allowVariantMatches">True when variant matches are allowed</param>
+        /// <returns>MethodDesc of the resolved virtual static method, null when not found (runtime lookup must be used)</returns>
+        private static MethodDesc ResolveVirtualStaticMethod(TypeDesc constrainedType, TypeDesc interfaceType, MethodDesc interfaceMethod, bool allowVariantMatches)
+        {
+            if (interfaceMethod.IsSharedByGenericInstantiations || interfaceType.IsCanonicalSubtype(CanonicalFormKind.Any))
+            {
+                return null;
+            }
+
+            // Check that there is no implementation of the interface on this type which is the canonical interface for a shared generic. If so, that indicates that
+            // we cannot exactly compute a target method result, as even if there is an exact match in the type hierarchy
+            // it isn't guaranteed that we will always find the right result, as we may find a match on a base type when we should find the match
+            // on a more derived type.
+            TypeDesc interfaceTypeCanonical = interfaceType.ConvertToCanonForm(CanonicalFormKind.Specific);
+            if (interfaceType != interfaceTypeCanonical)
+            {
+                foreach (DefType runtimeInterface in constrainedType.RuntimeInterfaces)
+                {
+                    if (interfaceTypeCanonical == runtimeInterface)
+                    {
+                        return null;
+                    }
+                }
+            }
+
```

Comment:
This logic that refers to Canon, and canonicalization is specific to actual compilation with utilizing a Canonical representation. I would either keep this in CorInfoImpl.ReadyToRun.cs or move move this to the DevirtualizationManager.

Reply:
Actually given some more thought, this should definitely be part of DevirtualizationManager, and we probably need to have a different result for the conditions of failure due to various canonical shenanigans, and failure due to the resolution algorithm failing.

Note, the canonicalization failure stuff should be in the Common version of DevirtualizationManager.

We should also take a very close look at the version resilience characteristics of all of this, and consider if there are scenarios where we need to abort the calculation, and fallback to a RequiresRuntimeJitException. (That logic would go in the R2R specific DevirtualizationManager.

The resolution algorithm failing should trigger a CodeGenerationException, or RequiresRuntimeJitException (and also probably an assert), and the canonical failure, should trigger the change to use CORINFO_CALL_POINTER.

Reply:
I have moved the core logic to DevirtualizationManaged and I have added initial provisions for version bubble checks. I think they aren't perfect but they're better than nothing. I'll work on further refining them.